packages/dashboard/src/app/views/myStories/filters/taxonomy/useTaxonomyFilters.js

```diff
@@ -0,0 +1,151 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import {
+  useCallback,
+  useEffect,
+  useState,
+  useMemo,
+} from '@googleforcreators/react';
+
+/**
+ * Internal dependencies
+ */
+import useApi from '../../../../api/useApi';
+
+export const getFilterPlaceholder = (filter) => {
+  return filter.labels?.allItems || filter.labels?.name || filter.restBase;
+};
+
+export const getFilterAriaLabel = (filter) => {
+  const ariaLable =
+    filter.labels?.singularName || filter.labels?.name || filter.restBase;
+  return `Filter stories by ${ariaLable}`;
+};
+
+export const getFilterSearchResultsLabel = (filter) => {
+  return (
+    filter.labels?.searchItems ||
+    (filter.labels?.name ? `Search ${filter.labels.name}` : 'Search Taxonomy')
+  );
+};
+
+/**
+ * Hook used for taxonomy filters logic
+ * Initilizes the taxonomy filters data
+ *
+ * @return {Array} [taxonomies, queryTaxonomyTerm] initial taxonomy filters data and a function to query the terms
+ */
+
+function useTaxonomyFilters() {
+  const [taxonomies, setTaxonomies] = useState([]);
+
+  const { getTaxonomies, getTaxonomyTerm } = useApi(
+    ({
+      actions: {
+        taxonomyApi: { getTaxonomies, getTaxonomyTerm },
+      },
+    }) => ({
+      getTaxonomies,
+      getTaxonomyTerm,
+    })
+  );
+
+  /**
+   * Query individual taxonomy data
+   * This is needed to initilize the primaryOptions and needed to search and set queriedOptions
+   *
+   * @param {Object} taxonomy object holding taxonomy data, including restBase and restPath used to fetch individual taxonomy terms
+   * @param {string|undefined} search string use to query taxonomy by name
+   * @return {Array} taxonomy terms
+   */
+  const queryTaxonomyTerm = useCallback(
+    async (taxonomy, search) => {
+      const { restBase, restPath } = taxonomy;
+      const terms = await getTaxonomyTerm(restPath, { search });
+      return terms.map((t) => ({
+        ...t,
+        restBase,
+        restPath,
+      }));
+    },
+    [getTaxonomyTerm]
+  );
+
+  /**
+   * Query all the taxonomies
+   * This should only be needed once get all the taxonomies and set individual term data
+   *
+   * @see queryTaxonomyTerm
+   */
+  const queryTaxonomies = useCallback(async () => {
+    const data = await getTaxonomies();
+    const promises = [];
+    const hierarchicalTaxonomies = data.filter(({ hierarchical }) =>
+      Boolean(hierarchical)
+    );
```

Comment:
@spacedmonkey WDYT about adding support for filtering hierarchical taxonomies in `Stories_Taxonomies_Controller::get_collection_params()`?

Then we could fetch `/wp-json/web-stories/v1/taxonomies?type=web-story&hierarchical=true` directly and don't need to do the filtering client-side.

Would be a nice follow-up code quality ticket.

Reply:
There are no functions to query by attributes and then get objects back. It would have to do manual searches for this prooperty. I dont see this being that big of a problem. This lookup is very quick.

Reply:
Also, It would be convenient to be able to fetch all the taxonomies with their terms instead of having to make multiple calls to the backend here, though Im not sure how difficult that would be to implement in.

Reply:
I did some research into this for now. There is no scalable way of doing this. If some taxonomies have 100+ terms, then that single request would result in far to many database queries.

Reply:
> There are no functions to query by attributes and then get objects back. It would have to do manual searches for this prowebperty. I dont see this being that big of a problem. This lookup is very quick.

There are. `get_taxonomies()` uses `wp_filter_object_list()` for that.

I'll whip up a quick PR to explain.

Reply:
See #11652

Reply:
> I did some research into this for now. There is no scalable way of doing this. If some taxonomies have 100+ terms, then that single request would result in far to many database queries.

+1

It's also not very RESTful